### ISSUE TITLE
Shard going 100% CPU 

### ISSUE THREAD

**Issue by geobeau** [scylladb/scylladb#16564]
This is Scylla's bug tracker, to be used for reporting bugs only.
If you have a question about Scylla, and not a bug, please ask it in
our mailing-list at scylladb-dev@googlegroups.com or in our slack channel.

- [x] I have read the disclaimer above, and I am reporting a suspected malfunction in Scylla.

*Installation details*
Scylla version (or git commit hash): 5.2.7
Cluster size: 16 (6 nodes in DC-A, 10 nodes in DC-B)
OS (RHEL/CentOS/Ubuntu/AWS AMI): Kubernetes (Centos image)

*Hardware details (for performance issues)* 
Platform (physical/VM/cloud instance type/docker): Docker
Hardware: sockets=1 cores=8 hyperthreading=16 (14 allocated to container) memory=256 (240 allocated to container)
Disks: (SSD/HDD, count) 6 SSD, Hardware RAID 0

In a cluster with many tables and materialized views, we have the following pattern repeating itself:

After an unknown event, one of the shard (of one of the node) will start to run at 100% and never recover by itself:
<img width="691" alt="Screenshot 2023-12-22 at 16 13 48" src="https://github.com/scylladb/scylladb/assets/12186336/468a4612-9d80-4c3f-aad6-9fdeca2ab36d">

The increase to 100% seems to be linked to the shard losing its cache and saturating the disks of the node:
<img width="1362" alt="Screenshot 2023-12-22 at 16 19 16" src="https://github.com/scylladb/scylladb/assets/12186336/820e29c7-64cd-4bae-9845-49f8da12584f">

<img width="690" alt="Screenshot 2023-12-22 at 16 14 06" src="https://github.com/scylladb/scylladb/assets/12186336/89c6fde0-424a-4c25-9f26-866dfa0ad119">

We can see a large eviction of cache on the shard impacted:
<img width="1353" alt="Screenshot 2023-12-22 at 16 15 55" src="https://github.com/scylladb/scylladb/assets/12186336/46ee6782-d39b-4838-a038-9c51116952bc">

I suspect it come from a change in the memory of scylla.

We can see a big eviction of LSA memory:

<img width="688" alt="Screenshot 2023-12-23 at 10 12 21" src="https://github.com/scylladb/scylladb/assets/12186336/a5d16839-4628-45b0-b4db-47a55980bf0d">

We can the LSA shrinking:
<img width="684" alt="Screenshot 2023-12-23 at 10 12 53" src="https://github.com/scylladb/scylladb/assets/12186336/86b85cae-a8cd-4ac5-bb28-0a6db74d73b3">

But still reporting free space (that could be used):
<img width="693" alt="Screenshot 2023-12-23 at 10 20 35" src="https://github.com/scylladb/scylladb/assets/12186336/b3e7e8ae-36a4-49ca-ad63-efa66fa68f4b">

Do you know what could trigger this cache eviction?

**Comment by mykaul** [comment]
Do you have any logs relevant to the time of such events?

**Comment by geobeau** [comment]
No, there are logs about too many tombstones: `querier - Read 259 live rows and 1145 tombstones for ...` and compactions but they are present all the time.

It seems that a working mitigation on our side is forcing a compaction on one of the table: `nodetool compact <keyspace> <table>`. Otherwise the shard can stay at 100% for days

**Comment by mykaul** [comment]
I assume (but don't know) that compaction is running - you should be able to see it in the monitoring as well. But it definitely should finish running. Without logs it'd be difficult to understand what's going on.

**Comment by avikivity** [comment]
Please use `perf top --cpu` on that shard's cpu.

**Comment by mykaul** [comment]
Perhaps, now that I re-read your description, you may be suffering from https://github.com/scylladb/scylladb/issues/15844 - which was fixed in 5.2.11.

**Comment by avikivity** [comment]
cache eviction is usually due to someone trying to allocate a large contiguous block of memory. The logs will show it as a large allocation warning with a stacktrace.

**Comment by geobeau** [comment]
> Please use perf top --cpu on that shard's cpu.

I will do it as soon as the issue is reproduced. In a previous occurrence I took a perf of the CPU that was slow (perf record -g -C <cpu>):
<img width="1188" alt="Screenshot 2023-12-07 at 22 02 56" src="https://github.com/scylladb/scylladb/assets/12186336/71811e6a-8904-4880-a0ed-413a4007e467">

> Perhaps, now that I re-read your description, you may be suffering from https://github.com/scylladb/scylladb/issues/15844 - which was fixed in 5.2.11.

From the descriptions, it seems that the cpu drops to 0% instead. I will try to bump to the latest

**Comment by geobeau** [comment]
This contains logs from 16:00 to 16:30 of the impacted instance. The impacted shard is `shard 3` and the issue starts at around 16:12.

[pod1.logs.zip](https://github.com/scylladb/scylladb/files/13772223/pod1.logs.zip)

**Comment by avikivity** [comment]
Looks like it's spending time on index entries. What's your cache_index_pages? recommend flipping it to `true` if not already and trying it out.

**Comment by avikivity** [comment]
@denesb 

>  (inlined by) evictable_reader_v2::fill_buffer() at ./readers/multishard.cc:612

Looks like a semi-bounded loop. It's bounded, but I'm not sure by what.

@geobeau do you have tons of range tombstones? and what was the size of the allocation reported?

In any case this looks like a minor issue with no real impact.

**Comment by geobeau** [comment]
I have upgraded the cluster to 5.2.12

> What's your cache_index_pages? recommend flipping it to true if not already and trying it out.
cache_index_pages was set to false (we add a performance issue on a different cluster). I am re-enabling it.

> do you have tons of range tombstones?
Yes, in the logs I attached we can see a lot of tombstones are scanned. We set a rather aggressive compaction strategy but there are still a lot of them.

> what was the size of the allocation reported?

```
December 26th 2023, 18:57:07.467﻿WARN  2023-12-26 18:57:07,466 [shard 10] seastar_memory - oversized allocation: 1048576 bytes. This is non-fatal, but could lead to latency and/or fragmentation issues. Please report: at 0x53d1dbe 0x53d2370 0x53d2658 0x5001bba 0x5003127 0x5005b81 0x1e5ec09 0x120d49a 0x5040d84 0x5042007 0x5062df1 0x50130ea /opt/scylladb/libreloc/libc.so.6+0x8b12c /opt/scylladb/libreloc/libc.so.6+0x10cbbf
December 26th 2023, 16:29:10.907﻿WARN  2023-12-26 16:29:10,907 [shard 13] seastar_memory - oversized allocation: 1048576 bytes. This is non-fatal, but could lead to latency and/or fragmentation issues. Please report: at 0x53d1dbe 0x53d2370 0x53d2658 0x5001bba 0x5003127 0x5005b81 0x1e5ec09 0x120d49a 0x5040d84 0x5042007 0x5062df1 0x50130ea /opt/scylladb/libreloc/libc.so.6+0x8b12c /opt/scylladb/libreloc/libc.so.6+0x10cbbf
December 23rd 2023, 06:44:21.966﻿WARN  2023-12-23 06:44:21,966 [shard 13] seastar_memory - oversized allocation: 1048576 bytes. This is non-fatal, but could lead to latency and/or fragmentation issues. Please report: at 0x53d1dbe 0x53d2370 0x53d2658 0x5001bba 0x5003127 0x5005b81 0x1e5ec09 0x120d49a 0x5040d84 0x5042007 0x5062df1 0x50130ea /opt/scylladb/libreloc/libc.so.6+0x8b12c /opt/scylladb/libreloc/libc.so.6+0x10cbbf
December 23rd 2023, 01:16:34.700﻿WARN  2023-12-23 01:16:34,699 [shard  4] seastar_memory - oversized allocation: 1048576 bytes. This is non-fatal, but could lead to latency and/or fragmentation issues. Please report: at 0x53d1dbe 0x53d2370 0x53d2658 0x5001bba 0x5003127 0x5005b81 0x1e5ec09 0x120d49a 0x5040d84 0x5042007 0x5062df1 0x50130ea /opt/scylladb/libreloc/libc.so.6+0x8b12c /opt/scylladb/libreloc/libc.so.6+0x10cbbf
December 23rd 2023, 01:16:34.166﻿WARN  2023-12-23 01:16:34,165 [shard  4] seastar_memory - oversized allocation: 1048576 bytes. This is non-fatal, but could lead to latency and/or fragmentation issues. Please report: at 0x53d1dbe 0x53d2370 0x53d2658 0x5001bba 0x5003127 0x5005b81 0x1e5ec09 0x120d49a 0x5040d84 0x5042007 0x5062df1 0x50130ea /opt/scylladb/libreloc/libc.so.6+0x8b12c /opt/scylladb/libreloc/libc.so.6+0x10cbbf
December 23rd 2023, 01:16:34.098﻿WARN  2023-12-23 01:16:34,098 [shard  4] seastar_memory - oversized allocation: 1048576 bytes. This is non-fatal, but could lead to latency and/or fragmentation issues. Please report: at 0x53d1dbe 0x53d2370 0x53d2658 0x5001bba 0x5003127 0x5005b81 0x1e5ec09 0x120d49a 0x5040d84 0x5042007 0x5062df1 0x50130ea /opt/scylladb/libreloc/libc.so.6+0x8b12c /opt/scylladb/libreloc/libc.so.6+0x10cbbf
December 22nd 2023, 03:29:12.763﻿WARN  2023-12-22 03:29:12,762 [shard 10] seastar_memory - oversized allocation: 1048576 bytes. This is non-fatal, but could lead to latency and/or fragmentation issues. Please report: at 0x53d1dbe 0x53d2370 0x53d2658 0x5001bba 0x5003127 0x5005b81 0x1e5ec09 0x120d49a 0x5040d84 0x5042007 0x5062df1 0x50130ea /opt/scylladb/libreloc/libc.so.6+0x8b12c /opt/scylladb/libreloc/libc.so.6+0x10cbbf
December 22nd 2023, 03:29:12.717﻿WARN  2023-12-22 03:29:12,717 [shard 10] seastar_memory - oversized allocation: 1048576 bytes. This is non-fatal, but could lead to latency and/or fragmentation issues. Please report: at 0x53d1dbe 0x53d2370 0x53d2658 0x5001bba 0x5003127 0x5005b81 0x1e5ec09 0x120d49a 0x5040d84 0x5042007 0x5062df1 0x50130ea /opt/scylladb/libreloc/libc.so.6+0x8b12c /opt/scylladb/libreloc/libc.so.6+0x10cbbf
December 22nd 2023, 03:29:12.671﻿WARN  2023-12-22 03:29:12,670 [shard 10] seastar_memory - oversized allocation: 1048576 bytes. This is non-fatal, but could lead to latency and/or fragmentation issues. Please report: at 0x53d1dbe 0x53d2370 0x53d2658 0x5001bba 0x5003127 0x5005b81 0x1e5ec09 0x120d49a 0x5040d84 0x5042007 0x5062df1 0x50130ea /opt/scylladb/libreloc/libc.so.6+0x8b12c /opt/scylladb/libreloc/libc.so.6+0x10cbbf
December 21st 2023, 17:50:06.241﻿WARN  2023-12-21 17:50:06,240 [shard  5] seastar_memory - oversized allocation: 1212416 bytes. This is non-fatal, but could lead to latency and/or fragmentation issues. Please report: at 0x53d1dbe 0x53d2370 0x53d2658 0x5001bba 0x5003127 0x5005b81 0x35cbe4a 0x1925596 0x5040d84 0x5042007 0x5062df1 0x50130ea /opt/scylladb/libreloc/libc.so.6+0x8b12c /opt/scylladb/libreloc/libc.so.6+0x10cbbf
```

**Comment by avikivity** [comment]
ok, we need to wait for @denesb to give an opinion here.

**Comment by geobeau** [comment]
Got another occurrence with 5.2.12 and index caching:

![Screenshot 2023-12-29 at 14 32 25](https://github.com/scylladb/scylladb/assets/12186336/3048d82d-eb4d-4107-b575-4bcab1cd9d31)

**Comment by geobeau** [comment]
Correction: it seems to have been transient. It showed the same pattern as before except this time it recovered after 10 minutes by itself. Before it never recovered.

<img width="692" alt="Screenshot 2023-12-29 at 15 52 17" src="https://github.com/scylladb/scylladb/assets/12186336/05454eef-ebed-4858-a6e0-9da5b5f5e08f">

I will update the issue if we see a long saturation come back

**Comment by geobeau** [comment]
We don't see shard being saturated for a long time anymore. But we have had nodes crashing during repair operations.

Here is the stack trace from the multiple crashes:
```
December 29th 2023, 21:44:10.378﻿WARN  2023-12-29 21:44:10,378 [shard  8] repair - repair_writer: keyspace=keyspace1, table=candidates_by_sku_uid_index, multishard_writer failed: std::bad_alloc (std::bad_alloc)December 29th 2023, 21:44:10.378﻿WARN  2023-12-29 21:44:10,378 [shard  8] repair - repair_writer: keyspace=keyspace1, table=candidates_by_sku_uid_index, wait_for_writer_done failed: std::bad_alloc (std::bad_alloc)
December 29th 2023, 21:44:10.639﻿WARN  2023-12-29 21:44:10,639 [shard  6] storage_proxy - Exception when communicating with 192.168.93.53, to read from keyspace1.candidates: Tombstones processed by unpaged query exceeds limit of 10000 (configured via query_tombstone_page_limit)
December 29th 2023, 21:44:10.707﻿WARN  2023-12-29 21:44:10,706 [shard  6] storage_proxy - Exception when communicating with 192.168.59.73, to read from keyspace1.candidates: Tombstones processed by unpaged query exceeds limit of 10000 (configured via query_tombstone_page_limit)
December 29th 2023, 21:44:10.790﻿WARN  2023-12-29 21:44:10,790 [shard  0] querier - Read 186 live rows and 11405 tombstones for keyspace1.candidates partition key "foobar" {{3846364136495163941, pk{002b00000004000000830000000e3520372061726561207275672078000000040000000200000005656e2d5553}}} (see tombstone_warn_threshold)
December 29th 2023, 21:44:10.791﻿ERROR 2023-12-29 21:44:10,790 [shard  6] storage_proxy - Exception when communicating with 192.168.56.47, to read from keyspace1.candidates: std::runtime_error (Tombstones processed by unpaged query exceeds limit of 10000 (configured via query_tombstone_page_limit))
December 29th 2023, 21:44:13.826﻿WARN  2023-12-29 21:44:13,826 [shard  4] repair - repair_writer: keyspace=keyspace1, table=candidates_by_sku_uid_index, multishard_writer failed: std::bad_alloc (std::bad_alloc)
December 29th 2023, 21:44:13.827﻿WARN  2023-12-29 21:44:13,826 [shard  4] repair - repair_writer: keyspace=keyspace1, table=candidates_by_sku_uid_index, wait_for_writer_done failed: std::bad_alloc (std::bad_alloc)
December 29th 2023, 21:44:14.099﻿ERROR 2023-12-29 21:44:14,099 [shard  9] storage_proxy - Exception when communicating with 192.168.56.47, to read from keyspace1.candidates: std::bad_alloc (std::bad_alloc)
December 29th 2023, 21:44:14.099﻿ERROR 2023-12-29 21:44:14,098 [shard  4] sstable - failed reading index for /var/lib/scylla/data/keyspace1/candidates-0e458390ecc211ecbca0000000000004/me-21704-big-Data.db: std::bad_alloc (std::bad_alloc)
 seastar::internal::coroutine_traits_base<seastar::foreign_ptr<seastar::lw_shared_ptr<nonwrapping_interval<dht::ring_position> const> > >::promise_type
0x5137772
0x5105438
0x1ea1525
0x5137c21
 seastar::internal::coroutine_traits_base<void>::promise_type
0x123836a
0x1e881c5
 --------
/opt/scylladb/libreloc/libc.so.6+0x267fb
 --------
ERROR 2023-12-29 21:44:14,100 [shard  4] mutation_reader - 14merging_readerI22mutation_reader_mergerE [0x6040074c0380]: permit keyspace1.candidates_by_sku_uid_index:shard-reader: was not closed before destruction, at: 0x54a765e 0x54a7c10 0x54a7ef8 0x50e7623 0x1ea150e 0x1e881c5 0x123836a 0x5115cb4 0x5116f37 0x5137c21 0x50e822a /opt/scylladb/libreloc/libc.so.6+0x8b19c /opt/scylladb/libreloc/libc.so.6+0x10cc5f
/opt/scylladb/libreloc/libc.so.6+0x8b19c
/opt/scylladb/libreloc/libc.so.6+0x3cab5
/opt/scylladb/libreloc/libc.so.6+0x3cb5f
0x5115cb4
Backtrace:
0x50e822a
/opt/scylladb/libreloc/libc.so.6+0x8cecb
/opt/scylladb/libreloc/libc.so.6+0x10cc5f
Aborting on shard 4.
 seastar::continuation<seastar::internal::promise_base_with_type<void>, seastar::smp_message_queue::async_work_item<(anonymous namespace)::shard_reader_v2::fast_forward_to(nonwrapping_interval<dht::ring_position> const&)::$_10>::run_and_dispose()::{lambda(auto:1)#1}, seastar::future<seastar::foreign_ptr<seastar::lw_shared_ptr<nonwrapping_interval<dht::ring_position> const> > >::then_wrapped_nrvo<void, seastar::smp_message_queue::async_work_item<(anonymous namespace)::shard_reader_v2::fast_forward_to(nonwrapping_interval<dht::ring_position> const&)::$_10>::run_and_dispose()::{lambda(auto:1)#1}>(seastar::smp_message_queue::async_work_item<(anonymous namespace)::shard_reader_v2::fast_forward_to(nonwrapping_interval<dht::ring_position> const&)::$_10>::run_and_dispose()::{lambda(auto:1)#1}&&)::{lambda(seastar::internal::promise_base_with_type<void>&&, seastar::smp_message_queue::async_work_item<(anonymous namespace)::shard_reader_v2::fast_forward_to(nonwrapping_interval<dht::ring_position> const&)::$_10>::run_and_dispose()::{lambda(auto:1)#1}&, seastar::future_state<seastar::foreign_ptr<seastar::lw_shared_ptr<nonwrapping_interval<dht::ring_position> const> > >&&)#1}, seastar::foreign_ptr<seastar::lw_shared_ptr<nonwrapping_interval<dht::ring_position> const> > >
0x5116f37
```

Decoded to:
```


seastar::print_with_backtrace(char const*, bool) at ./build/release/seastar/./seastar/src/core/reactor.cc:825
 (inlined by) seastar::sigabrt_action() at ./build/release/seastar/./seastar/src/core/reactor.cc:3864
 (inlined by) seastar::install_oneshot_signal_handler<6, (void (*)())(&seastar::sigabrt_action)>()::{lambda(int, siginfo_t*, void*)#1}::operator()(int, siginfo_t*, void*) const at ./build/release/seastar/./seastar/src/core/reactor.cc:3840
 (inlined by) __invoke at ./build/release/seastar/./seastar/src/core/reactor.cc:3836
void seastar::backtrace<seastar::backtrace_buffer::append_backtrace()::{lambda(seastar::frame)#1}>(seastar::backtrace_buffer::append_backtrace()::{lambda(seastar::frame)#1}&&) at ./build/release/seastar/./seastar/include/seastar/util/backtrace.hh:59
 (inlined by) ?? at ./build/release/seastar/./seastar/src/core/reactor.cc:783
 (inlined by) seastar::print_with_backtrace(seastar::backtrace_buffer&, bool) at ./build/release/seastar/./seastar/src/core/reactor.cc:813
flat_mutation_reader_v2::~flat_mutation_reader_v2() at ./readers/mutation_reader.cc:384
seastar::smp::configure(seastar::smp_options const&, seastar::reactor_options const&)::$_94::operator()() const at ./build/release/seastar/./seastar/src/core/reactor.cc:4326
 (inlined by) void std::__invoke_impl<void, seastar::smp::configure(seastar::smp_options const&, seastar::reactor_options const&)::$_94&>(std::__invoke_other, seastar::smp::configure(seastar::smp_options const&, seastar::reactor_options const&)::$_94&) at /usr/bin/../lib/gcc/x86_64-redhat-linux/12/../../../../include/c++/12/bits/invoke.h:61
 (inlined by) std::enable_if<is_invocable_r_v<void, seastar::smp::configure(seastar::smp_options const&, seastar::reactor_options const&)::$_94&>, void>::type std::__invoke_r<void, seastar::smp::configure(seastar::smp_options const&, seastar::reactor_options const&)::$_94&>(seastar::smp::configure(seastar::smp_options const&, seastar::reactor_options const&)::$_94&) at /usr/bin/../lib/gcc/x86_64-redhat-linux/12/../../../../include/c++/12/bits/invoke.h:111
 (inlined by) std::_Function_handler<void (), seastar::smp::configure(seastar::smp_options const&, seastar::reactor_options const&)::$_94>::_M_invoke(std::_Any_data const&) at /usr/bin/../lib/gcc/x86_64-redhat-linux/12/../../../../include/c++/12/bits/std_function.h:290

seastar::internal::coroutine_traits_base<void>::promise_type[Backtrace #1]
std::_Sp_counted_base<(__gnu_cxx::_Lock_policy)2>::_M_release_last_use_cold() at /usr/bin/../lib/gcc/x86_64-redhat-linux/12/../../../../include/c++/12/coroutine:244
 (inlined by) seastar::internal::coroutine_traits_base<void>::promise_type::run_and_dispose() at ././seastar/include/seastar/core/coroutine.hh:120
seastar::optimized_optional<flat_mutation_reader_v2>::~optimized_optional() at ././seastar/include/seastar/util/optimized_optional.hh:49
 (inlined by) evictable_reader_v2::fast_forward_to(nonwrapping_interval<dht::ring_position> const&) at ./readers/multishard.cc:651
/opt/scylladb/libreloc/libc.so.6: ELF 64-bit LSB shared object, x86-64, version 1 (GNU/Linux), dynamically linked, interpreter /lib64/ld-linux-x86-64.so.2, BuildID[sha1]=0dc6d3e329f8bf5e8c1de63c4c9d560fb9953ade, for GNU/Linux 3.2.0, stripped

addr2line: /opt/scylladb/libreloc/libc.so.6: unknown type [0x13] section `.relr.dyn'
abort at ??:?

seastar::internal::coroutine_traits_base<void>::promise_typeERROR 2023-12-29 21:44:14,100 [shard  4] mutation_reader - 14merging_readerI22mutation_reader_mergerE [0x6040074c0380]: permit rm_delivery.candidates_by_sku_uid_index:shard-reader: was not closed before destruction, at:
[Backtrace #2]
void seastar::backtrace<seastar::current_backtrace_tasklocal()::$_3>(seastar::current_backtrace_tasklocal()::$_3&&) at ./build/release/seastar/./seastar/include/seastar/util/backtrace.hh:59
 (inlined by) seastar::current_backtrace_tasklocal() at ./build/release/seastar/./seastar/src/util/backtrace.cc:86
seastar::current_tasktrace() at ./build/release/seastar/./seastar/src/util/backtrace.cc:137
seastar::current_backtrace() at ./build/release/seastar/./seastar/src/util/backtrace.cc:170
void log_error_and_backtrace<std::basic_string_view<char, std::char_traits<char> > >(seastar::logger&, std::basic_string_view<char, std::char_traits<char> > const&) at ./build/release/seastar/./seastar/src/core/on_internal_error.cc:38
 (inlined by) seastar::on_internal_error_noexcept(seastar::logger&, std::basic_string_view<char, std::char_traits<char> >) at ./build/release/seastar/./seastar/src/core/on_internal_error.cc:60
flat_mutation_reader_v2::~flat_mutation_reader_v2() at ./readers/mutation_reader.cc:383
seastar::optimized_optional<flat_mutation_reader_v2>::~optimized_optional() at ././seastar/include/seastar/util/optimized_optional.hh:49
 (inlined by) evictable_reader_v2::fast_forward_to(nonwrapping_interval<dht::ring_position> const&) at ./readers/multishard.cc:651
std::_Sp_counted_base<(__gnu_cxx::_Lock_policy)2>::_M_release_last_use_cold() at /usr/bin/../lib/gcc/x86_64-redhat-linux/12/../../../../include/c++/12/coroutine:244
 (inlined by) seastar::internal::coroutine_traits_base<void>::promise_type::run_and_dispose() at ././seastar/include/seastar/core/coroutine.hh:120
std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >::reserve(unsigned long) at ./build/release/seastar/./seastar/src/core/reactor.cc:2509
 (inlined by) seastar::reactor::run_some_tasks() at ./build/release/seastar/./seastar/src/core/reactor.cc:2946
seastar::reactor::do_run() at ./build/release/seastar/./seastar/src/core/reactor.cc:3115
seastar::smp::configure(seastar::smp_options const&, seastar::reactor_options const&)::$_94::operator()() const at ./build/release/seastar/./seastar/src/core/reactor.cc:4326
 (inlined by) void std::__invoke_impl<void, seastar::smp::configure(seastar::smp_options const&, seastar::reactor_options const&)::$_94&>(std::__invoke_other, seastar::smp::configure(seastar::smp_options const&, seastar::reactor_options const&)::$_94&) at /usr/bin/../lib/gcc/x86_64-redhat-linux/12/../../../../include/c++/12/bits/invoke.h:61
 (inlined by) std::enable_if<is_invocable_r_v<void, seastar::smp::configure(seastar::smp_options const&, seastar::reactor_options const&)::$_94&>, void>::type std::__invoke_r<void, seastar::smp::configure(seastar::smp_options const&, seastar::reactor_options const&)::$_94&>(seastar::smp::configure(seastar::smp_options const&, seastar::reactor_options const&)::$_94&) at /usr/bin/../lib/gcc/x86_64-redhat-linux/12/../../../../include/c++/12/bits/invoke.h:111
 (inlined by) std::_Function_handler<void (), seastar::smp::configure(seastar::smp_options const&, seastar::reactor_options const&)::$_94>::_M_invoke(std::_Any_data const&) at /usr/bin/../lib/gcc/x86_64-redhat-linux/12/../../../../include/c++/12/bits/std_function.h:290
std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >::_M_assign(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&) at /usr/bin/../lib/gcc/x86_64-redhat-linux/12/../../../../include/c++/12/bits/std_function.h:591
 (inlined by) seastar::posix_thread::start_routine(void*) at ./build/release/seastar/./seastar/src/core/posix.cc:73
pthread_condattr_setpshared at ??:?
__xmknodat at ??:?

seastar::internal::coroutine_traits_base<void>::promise_type[Backtrace #3]
pthread_condattr_setpshared at ??:?
gsignal at ??:?
__sigaction at ??:?
std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >::reserve(unsigned long) at ./build/release/seastar/./seastar/src/core/reactor.cc:2509
 (inlined by) seastar::reactor::run_some_tasks() at ./build/release/seastar/./seastar/src/core/reactor.cc:2946

Backtrace:[Backtrace #4]
std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >::_M_assign(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&) at /usr/bin/../lib/gcc/x86_64-redhat-linux/12/../../../../include/c++/12/bits/std_function.h:591
 (inlined by) seastar::posix_thread::start_routine(void*) at ./build/release/seastar/./seastar/src/core/posix.cc:73
pthread_key_delete at ??:?
__xmknodat at ??:?

seastar::continuation<seastar::internal::promise_base_with_type<void>, seastar::smp_message_queue::async_work_item<(anonymous namespace)::shard_reader_v2::fast_forward_to(nonwrapping_interval<dht::ring_position> const&)::$_10>::run_and_dispose()::{lambda(auto:1)#1}, seastar::future<seastar::foreign_ptr<seastar::lw_shared_ptr<nonwrapping_interval<dht::ring_position> const> > >::then_wrapped_nrvo<void, seastar::smp_message_queue::async_work_item<(anonymous namespace)::shard_reader_v2::fast_forward_to(nonwrapping_interval<dht::ring_position> const&)::$_10>::run_and_dispose()::{lambda(auto:1)#1}>(seastar::smp_message_queue::async_work_item<(anonymous namespace)::shard_reader_v2::fast_forward_to(nonwrapping_interval<dht::ring_position> const&)::$_10>::run_and_dispose()::{lambda(auto:1)#1}&&)::{lambda(seastar::internal::promise_base_with_type<void>&&, seastar::smp_message_queue::async_work_item<(anonymous namespace)::shard_reader_v2::fast_forward_to(nonwrapping_interval<dht::ring_position> const&)::$_10>::run_and_dispose()::{lambda(auto:1)#1}&, seastar::future_state<seastar::foreign_ptr<seastar::lw_shared_ptr<nonwrapping_interval<dht::ring_position> const> > >&&)#1}, seastar::foreign_ptr<seastar::lw_shared_ptr<nonwrapping_interval<dht::ring_position> const> > >[Backtrace #5]
seastar::reactor::do_run() at ./build/release/seastar/./seastar/src/core/reactor.cc:3115
```

Unfortunately, we were not able to retrieve a coredump (but we should retrieve it at next occurence)

**Comment by mykaul** [comment]
Split the above crash issues to https://github.com/scylladb/scylladb/issues/16606

**Comment by mykaul** [comment]
While the crash itself is interesting and needs to be looked at, you also have bad_alloc - do you have additional, earlier logs that we might be able to understand what's going on there?

**Comment by geobeau** [comment]
I have attached more logs from before the incident. We have addressed the logs with tombstones by reducing the GC interval.
(I have removed the name of the keys from logs) 
[logs.txt](https://github.com/scylladb/scylladb/files/13820855/logs.txt)

**Comment by avikivity** [comment]
You can consider tombstone_gc_mode = repair and repairing more frequently. This will minimize tombstones while avoiding resurrection risk.

**Comment by avikivity** [comment]
No you're using 5.2. There was a follow-up that improved things further, let me look.

**Comment by avikivity** [comment]
Maybe 854188a486142be4dd9ba51c388bdc9add69fc17.

**Comment by geobeau** [comment]
> You can consider tombstone_gc_mode = repair and repairing more frequently. This will minimize tombstones while avoiding resurrection risk.

Unfortunately, repair is creating a lot of load on the cluster so we throttle to run in ~24h and we run it every 7 days. So it would lead to accumulation of plenty of tombstones.

**Comment by avikivity** [comment]
Is your I/O scheduler configure correctly? It should be limited to 20% of the load.

**Comment by geobeau** [comment]
It is configured with the tuning tool provided. Looking at current metrics, it doesn't seem to affect performances significantly. Once we have situation under control I will check if we can run it more aggressively.

I have 2 more questions:
- repairs produces a big number of `repair_rx_row_nr` `repair_tx_row_nr`. I expect that under normal conditions we should have near 0 rows exchanged?
- Are there metrics to see the size of memtables in memory?

Thanks for the help 👍

**Comment by avikivity** [comment]
> It is configured with the tuning tool provided (we are using the k8s operator). Looking at current metrics, it doesn't seem to affect performances significantly. Once we have situation under control I will check if we can run it more aggressively.
> 
> I have 2 more questions:
> 
> * repairs produces a big number of `repair_rx_row_nr` `repair_tx_row_nr`. I expect that under normal conditions we should have near 0 rows exchanged?

How big?

Under a write load you can see some, I don't think repair is operating on a snapshot.

I think it's cumulative - it shows the count for all repairs since the instance started.

> * Are there metrics to see the size of memtables in memory?

scylla_memory_dirty_bytes - total memtable memory
scylla_memory_regular_unspooled_dirty_bytes - memtable memory that has not yet been flushed to an (unsealed) sstable


> 
> Thanks for the help 👍

**Comment by denesb** [comment]
We had a bug where the loop in the evictable reader's fill_buffer wouldn't stop, but the fix is alread in 5.2.7 (8e63b2f3e3).

**Comment by denesb** [comment]
@geobeau I just spotted this in the logs above:
```
December 29th 2023, 21:44:10.791﻿ERROR 2023-12-29 21:44:10,790 [shard  6] storage_proxy - Exception when communicating with 192.168.56.47, to read from keyspace1.candidates: std::runtime_error (Tombstones processed by unpaged query exceeds limit of 10000 (configured via query_tombstone_page_limit))
```
This might be (one of) the root of your problems: you have unpaged queries. They can consume a lot of memory. We have a safety check to kill them when they go way overboard, but still, they will do already do harm before they get there.

**Comment by denesb** [comment]
> @denesb isn't this fixed by [e9cbc9e](https://github.com/scylladb/scylladb/commit/e9cbc9ee85c25deb5ba9ee67ffa6e3ca9f904660)?

This cannot help when queries are unpaged.

**Comment by mykaul** [comment]
@geobeau - great to hear - what did you change for more aggressive compaction of the tombstones?
Can you quantify how much each change helped?

**Comment by geobeau** [comment]
The main thing was to reduce `gc_grace_seconds` drastically. As Avi mentioned it increase chances of resurrection but it's a tradeoff we are willing to take. We might review it again in the future.
Additionally, we made the sizedTiered compaction strategy able to compact more easily (min_threshold=2 and able to merge with bigger sstables). This had a marginal effect.

On percentage:
- running 5.2.12: 5% (it helped make sure we had all the fixes)
- using index caching: 15% (prevented the shard 100% and let the cluster show more symptoms of struggle with tombstones)
- aggressive compaction of tombstones: 80% (didn't see an issue since we changed the gc)

Again, thanks for the help
